src/diagnostics.rs

```diff
@@ -394,3 +395,507 @@ pub fn format_related_information(
                     .join("\n")
         })
 }
+
+// Pull diagnostics support (LSP 3.17+)
+pub fn text_document_diagnostic(meta: EditorMeta, ctx: &mut Context) {
+    let req_params: HashMap<_, _> = ctx
+        .servers(&meta)
+        .filter_map(|(server_id, server)| {
+            if !crate::capabilities::server_has_capability(
+                ctx.to_editor(),
+                server,
+                crate::capabilities::CAPABILITY_DIAGNOSTIC,
+            ) {
+                return None;
+            }
+
+            let buffile = &meta.buffile;
+            let uri = Url::from_file_path(buffile).unwrap();
+
+            let params = DocumentDiagnosticParams {
+                text_document: TextDocumentIdentifier { uri },
+                identifier: None,
+                previous_result_id: ctx
+                    .diagnostic_result_ids
+                    .get(&(server_id, buffile.clone()))
+                    .cloned(),
+                work_done_progress_params: WorkDoneProgressParams::default(),
+                partial_result_params: PartialResultParams::default(),
+            };
+
+            Some((server_id, vec![params]))
+        })
+        .collect();
+
+    if req_params.is_empty() {
+        return;
+    }
+
+    ctx.call::<DocumentDiagnosticRequest, _>(
+        meta,
+        RequestParams::Each(req_params),
+        move |ctx, meta, results| {
+            for (server_id, report) in results {
+                handle_document_diagnostic_response(server_id, report, &meta, ctx);
```

Comment:
Hmm I'm not sure if it's a good idea to run one `ctx.exec()` per server.
It doesn't really matter in practice as multi-servers is an edge case,
but ideally we coalesce them all in one big command.